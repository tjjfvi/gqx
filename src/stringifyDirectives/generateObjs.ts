
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
