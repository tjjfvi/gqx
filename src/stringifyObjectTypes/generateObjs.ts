
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
