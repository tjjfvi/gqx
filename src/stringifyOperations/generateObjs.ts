
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
