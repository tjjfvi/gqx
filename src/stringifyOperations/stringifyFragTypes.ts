
import { Obj, Context } from ".";
import stringifyNamespace from "./stringifyNamespace";

export default (objs: Obj[], ctx: Context) =>
  stringifyNamespace("$$Frag", objs, (o, p) =>
    `type ${p.id.prop} = $$DeepArray<${(p.type in ctx.objectTypes ? `${p.type}$` : "never")}>;`
  )
