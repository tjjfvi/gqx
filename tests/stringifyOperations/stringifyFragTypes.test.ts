import stringifyFragTypes from "../../src/stringifyOperations/stringifyFragTypes"
import generateObjs from "../../src/stringifyOperations/generateObjs"

test("", () => {
  let [objs, ctx] = generateObjs(utils.getOperations(["extensive"]));
  let code = stringifyFragTypes(objs, ctx);
  expect(code).toParse().toMatchSnapshot();
})
