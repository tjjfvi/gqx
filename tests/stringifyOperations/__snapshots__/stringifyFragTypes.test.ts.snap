// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"export namespace $$Frag {
  export namespace Query {
    export type getAuthor = Author$;
    export type getBook = Book$;
    export type listBooks = Book$;
  }
  export namespace Mutation {
    export type addReview = Review$;
  }
  export type $<T> =
    T extends typeof Query$getAuthor ? Query.getAuthor :
    T extends typeof Query$getBook ? Query.getBook :
    T extends typeof Query$listBooks ? Query.listBooks :
    T extends typeof Mutation$addReview ? Mutation.addReview :
    never
}"
`;
